### ISSUE TITLE
Clock should re-sort pending calls when one of them is reset

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4823]
|<img alt="alecu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| alecu reported|
|-|-|
|Trac ID|trac#4823|
|Type|defect|
|Created|2011-01-25 20:25:47Z|
|Branch|https://github.com/twisted/twisted/tree/keep-calls-ordered-4823|

twisted.internet.task.Clock sorts the list of pending calls only when a new call is added with callLater. If one of this pending calls is .reset() to a later time, calls scheduled to happen after the first will happen at the wrong point.

I'm attaching a testcase that highlight this issue.

Attachments:

* [test_clocksorting.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b3e/b3eefa921a7af008fe14fa2dd88d5a9073cbf9ba/f227e858203c3e374ecbb25a9f9abd20ce755cda.py) (949 bytes) - added by alecu on 2011-01-25 20:27:08Z - 
* [keep-calls-ordered.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b3e/b3eefa921a7af008fe14fa2dd88d5a9073cbf9ba/56bf3a80fb0c129e74e379114d16f5ccef1e5cb7.patch) (1932 bytes) - added by alecu on 2011-01-25 20:47:21Z - 
* [keep-calls-ordered-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/b3e/b3eefa921a7af008fe14fa2dd88d5a9073cbf9ba/992c3db9a0bd27421ee22b09035f6e4a9c28a6fa.patch) (3553 bytes) - added by alecu on 2011-02-08 18:03:27Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4823 4823
type__defect defect
reporter__alecu alecu
priority__normal normal
milestone__ 
branch__branches_keep_calls_ordered_4823 branches/keep-calls-ordered-4823
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1295987147000000 1295987147000000
changetime__1297650176000000 1297650176000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_1"></a>|@jyknight commented|
|-|-|

Sigh. When I saw this at first I thought you were saying the reactor was broken like that (which would've been quite surprising). But now I see that task.Clock has its own broken re-implementation of callLater. It should really use the same implementation as the reactor, just with a different "self.seconds()" method.

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)<a name="note_2"></a>|@jyknight commented|
|-|-|

(that was not a review of your patch, just a lament.) BTW, when you attach a patch proposed to fix an issue, you should add the "review" keyword so that someone will review it.

**Comment by twisted-trac** [comment]
|<img alt="alecu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|alecu commented|
|-|-|

Hi all, I'd like some reviews on the keep-calls-ordered.patch.

This patch makes sure that pending calls are re-sorted when they are about to be processed (and not only when they are added, because one of them might have been rescheduled.)

**Comment by twisted-trac** [comment]
|<img alt="alecu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|alecu commented|
|-|-|

(didn't notice you removed the "clock" keyword, sorry)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to alecu|
|-|-|

Hum, it would be nice to use the reactor semantics a bit more, but I guess it's hard. In the mean time, I've noticed a small problem not fixed by your solution, if you call reset in the callLater itself:

```
        result = []
        expected = [('c', 3.0)]
        clock = task.Clock()
        logtime = lambda n: result.append((n, clock.seconds()))
        
        def a():
            call_b.reset(3.0)
        
        call_a = clock.callLater(1.0, a)
        call_b = clock.callLater(2.0, logtime, "b")
        call_c = clock.callLater(3.0, logtime, "c")
        
        clock.advance(3)
        self.assertEqual(result, expected)
```

I think the fix is to call sort again after call.func in advance. As it will be the third call of call.sort, it will probably worth extracting it in a method.

Also, please name your tests 'test_callLaterX', and add a NEWS fragment. Thanks!

**Comment by twisted-trac** [comment]
|<img alt="alecu's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|alecu set owner to @therve|
|-|-|

Hi Thomas, thanks for the review. I'm attaching another a second diff with the fixes you suggested.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

alecu, in the future, when submitting tickets for review you should generally leave them un-assigned, since an individual reviewer might not have time to do all the subsequent reviews.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

(In [30722]) Branching to 'keep-calls-ordered-4823'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph commented|
|-|-|

[Build Results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/keep-calls-ordered-4823)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

(In [30734]) review trivia 1: add whitespace (2 blank lines between methods, 3 between classes) refs [#4823](https://github.com/twisted/twisted/issues/4823)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

(In [30735]) review trivia 2: remove assigned-but-not-used variables so pyflakes won't complain.  refs [#4823](https://github.com/twisted/twisted/issues/4823)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph commented|
|-|-|

(In [30738]) review trivia 3: improve test docstrings, and make  slightly more robust / pleasant.  (don't refer to locals in functions defined before those locals are assigned.) refs [#4823](https://github.com/twisted/twisted/issues/4823)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

(In [30739]) review trivia 4: tweak implementatino whitespace) refs [#4823](https://github.com/twisted/twisted/issues/4823)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

Just re-ran the build to make sure my trivial changes didn't actually touch anything meaningful.  Will be committing momentarily.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph set status to `closed`|
|-|-|

(In [30741]) twisted.internet.task.Clock now sorts the list of pending calls before and after processing each call.

Author: alecu

Reviewer: glyph

Fixes: [#4823](https://github.com/twisted/twisted/issues/4823)

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
